---
title: "Analysis of 2-4-8 study data"
output: html_notebook
---

## Galaxy workflow 

The following FASTQ datasets were analyzed with a modified version of the
AMRPlusPlus workflow in the Galaxy instance of the
National Microbiology Laboratory (Public Health Agency of Canada). 

1. Forward and reverse reads that were merged from different lanes (full)

2. Forward and reverse FASTQ files of the individual datasets (half)

3. Forward and reverse with half of the individual datasets (quarter)

4. Subsampled sequences with exactly half of individual datasets (quarter-seqtk)

The modifications to the AMRPlusPlus include the integration of taxonomical
classification (Kraken) in the same workflow, and the ommission of filtering
reads vs. host genome.

### Taxonomic classification

Trimmed reads were classified with `kraken` using a database containing
bacterial, virual, fungal, protozoal and archaeal genomes. Classified reads were
then filtered with `kraken-filter` (score = 0.05). Finally, a report of the
taxonomic classification was generated with `kraken-report`.

## Exploratory and statistical analyses

### Venn diagrams

__AMR__: AMR reports resulting from Coverage Sampler analyses were processed by 
filtering out those results with a gene fraction less than 75%, and also by 
filtering out housekeeping genes (genes present in the MEGARes database which 
require SNP confirmation).

Table 1. Genes removed from the AMR Coverage Sampler dataset

__Microbiome__: 

Results for all Eukaryotes and for Enterobacteria phage phiX174 sensu lato were 
filtered out from the Kraken reports with Pavian. Additionally, taxonomic
lineages were added with Pavian.

Sets of unique taxa at each level among phylum, class, family, order, genus, and
species were determined using Python 3.4.2. Venn diagrams were plotted using the
`matplotlib_venn` package in Python 3.4.2.

### Generation of rarefaction curves

__AMR__: Rarefaction curves were generated using two methods:

1. Rarefaction curves were built with the results from Rarefaction Analyzer 
based on __percent__ of the SAM file sampled. The mean of the counts was 
calculated, and the standard error was plotted.

2. Rarefaction curves were also generated by using the `vegan` package of R
based on the __number__ of reads sampled.

2. In the case of Kraken data, the rarefaction curves were built with the 
results from processing the reads with the `vegan` package of R.

    * Plot with `ggplot2`
  
  |Sample|Sampling Size|No. of OTUs|
  |------|-------------|-----------|
  

4. Processing with Steven's scripts
