---
title: "Analysis of 2-4-8 study data"
output: html_notebook
---

## Galaxy workflow 

The following FASTQ datasets were analyzed with a modified version of the
AMRPlusPlus workflow in the Galaxy instance of the
National Microbiology Laboratory (Public Health Agency of Canada). 

1. Forward and reverse reads that were merged from different lanes (full)

2. Forward and reverse FASTQ files of the individual datasets (half)

3. Forward and reverse with half of the individual datasets (quarter)

4. Subsampled sequences with exactly half of individual datasets (quarter-seqtk)


The modifications to the AMRPlusPlus include the integration of taxonomical
classification (Kraken) in the same workflow, and the ommission of filtering
reads vs. host genome.

The Galaxy workflow includes the following processes:

1. __Trimming and quality assessment__
2. __AMR classification__
3. __Taxonomic classification__

In this document, the statistical analyses will also be discussed.

### Trimming and quality assessment

Reads were trimmed with Trimmomatic, using the following paramaters:
    * ILLUMINACLIP trimming searching for TruSeq3 paired-end adapters 
    (2:30:10:3)
    * Trimmed the leading 3 bases
    * Trimmed the trailing 3 bases
    * Sliding window quality filtering: every four bases, minimum quality= 15
    * Kept reads with a length of at least 36 bases after trimming
  
### AMR classification

The reads resulting after trimming were mapped vs. the MEGAREs database v1.01
using the Map with BWA tool in (version 0.7.12.1). Then, the resulting BAM
alignment was sorted by coordinates using the X tool, and and then it was
converted to SAM format. The SAM file was analyzed with 3 tools: Coverage
Sampler, SnipFinder, RarefactionAnalyzer


### Taxonomic classification

Trimmed reads were classified with `kraken` using a database containing
bacterial, virual, fungal, protozoal and archaeal genomes. Classified reads were
then filtered with `kraken-filter` (score = 20). Finally, a report of the
taxonomic classification was generated with `kraken-report`.

## Exploratory and statistical analyses

AMR reports resulting from Coverage Sampler analyses were further processed to add a column with sample name, depth level, and alignment method. Then, all the reports were combined into a single document that was further processed with R.

Kraken reports were read into R using the `readr` package, whitespace was removed, and all reports were combined into a single script.

### Generation of rarefaction curves

1. Reshape data to wide format

2. Normalize data with cumSum normalization

3. Removal of housekeeping genes (AMR) and PhiX (Kraken)

4. Construction of rarefaction curves. In the case of AMRPlusPlus data, the rarefaction curves were built with the results from Rarefaction Analyzer. In the case of Kraken data, the rarefaction curves were built with the results from processing the reads with the `vegan` package of R.

    * Plot with `ggplot2`
  
  |Sample|Sampling Size|No. of OTUs|
  |------|-------------|-----------|
  

4. Processing with Steven's scripts
