---
title: "2-4-8 study: microbiome rarefaction curves, alpha-diversity, and richness"
output: 
  word_document: default
  html_notebook: default
bibliography: 2-4-8_library.bib
---

# Methods

## Rarefaction curves of microbiome results

Rarefaction curves were generated by using the `rarecurve` function of 
the `vegan` package of R [@vegan_oksanen_2017]. In order to generate the rarefaction
curves, Kraken results were split by taxonomic level, and subsamples were taken at 
every 1,000 hits using a sampling size equal to the minimum sum of Kraken hits among
the 32 samples analyzed in this study. The results for all taxonomic levels were 
combined into a single data frame, and only the data for Phylum, Order, Class, 
Family, Genus, and Species were used for visualization and further analysis. The 
rarefaction curves were plotted using the `ggplot2` package of R 
[@wickham_ggplot2_2009]. Facet grids were generated using the Depth category.

## Alpha diversity and species richness of microbiome results

Alpha-diversity was calculated as the Inverse Simpson's index for rarefied data on 
every sample using the `diversity` function of `vegan` with the `index=invsimpson` 
parameter. The distributions of alpha-diversity were plotted for every taxonomic 
level and for each sequencing depth level as boxplots using `ggplot2`.

Species richness was calculated for every sample using the `specnumber` function of 
`vegan`. The distributions of species richness for raw data were visualized as 
boxplots in the same manner as alpha-diversity

# References

